### PR TITLE
Revamping block validation interface interaction with wallet

### PR DESCRIPTION
**Pull request by furszy**
Revamped the validation interface interaction with the wallet, encapsulating and improving the mempool and block signaling and each of the wallet signals handler.
Restructured the Sapling witnesses and nullifiers update under the new signals.
Solved several bugs that found on the way as well (Check each commit description).

This PR concludes with #1726 long road :). Pushing our repository around 2 years ahead in btc time in the mempool and validation interface areas (without including RBF). 
The new validation -> wallet interaction architecture will enable further, and much more user facing important, improvements for the syncing process, overall software responsiveness and multithreading locking issues.

Adapted backports:
#6464 --> Always clean up manual transaction prioritization (mempool)
#9240 --> Remove txConflicted (mempool)
#9371 --> Notify on removal (mempool) 
#9605 --> Use CScheduler for wallet flushing, remove ThreadFlushWalletDB (walletdb)
#9725 --> CValidationInterface Cleanups (wallet, validation and validation interface)

### CODE REVIEW

**Comment by furszy** [comment]
Rebased on master due #2118 and #2126 pre-requisites merge. 
Ready for review and finally kill #1726 :).

**Review by random-zebra — COMMENTED** [review]
Looking pretty good. Concept/code ACK. Will do some testing.

Seems that the backport missed 73296f54d6f0b38fcac627882d2e9480a2bad14a, so the wallet flush now is scheduled every 500 seconds, instead of half second (previously the `ThreadFlushWalletDB` had `MilliSleep(500)`).

**Comment by furszy** [comment]
yeah zebra, the reason why 73296f5 isn't here is because i did it in #1973 some time ago and was expecting to merge that one before #1726 long road but.. got pretty obnubilated finishing all of this hehe. Going to cherry-pick it here, good catch 👍 .

File: src/validation.cpp [hunk 1]
```diff
@@ -186,10 +186,11 @@ class MemPoolConflictRemovalTracker
 private:
     std::vector<CTransactionRef> conflictedTxs;
     CTxMemPool &pool;
+    std::unique_ptr<interfaces::Handler> m_handler_notify_entry_removed;
 
 public:
     MemPoolConflictRemovalTracker(CTxMemPool &_pool) : pool(_pool) {
-        pool.NotifyEntryRemoved.connect(std::bind(&MemPoolConflictRemovalTracker::NotifyEntryRemoved, this, std::placeholders::_1, std::placeholders::_2));
+        m_handler_notify_entry_removed = interfaces::MakeHandler(pool.NotifyEntryRemoved.connect(std::bind(&MemPoolConflictRemovalTracker::NotifyEntryRemoved, this, std::placeholders::_1, std::placeholders::_2)));
```
[the remark this answers is not in the document]
Comment (furszy) [reply src/validation.cpp]: aren't removed, are moved into `ConnectTrace` same as the conflicted transactions vector. Adding the changes here or in a commit post fcd4446, code wise, is the same. The difference relies on the compilation, which wouldn't work until that commit.

This was done to adapt the code to our sources. In upstream's PR, the signal bind and disconnect is done via boost::bind which we can't have as every boost bind usage was removed from the sources in #2082, replacing it by a generic event handler.

**Comment by furszy** [comment]
Feedback tackled, added 4bdf68a solving the wallet flushing time and the nullptr and extra argument nits.

File: src/scheduler.h [hunk 2]
```diff
@@ -45,14 +45,14 @@ class CScheduler
...
-    void scheduleFromNow(Function f, int64_t deltaSeconds);
+    void scheduleFromNow(Function f, int64_t deltaMilliSeconds);
 
     // Another convenience method: call f approximately
     // every deltaSeconds forever, starting deltaSeconds from now.
     // To be more precise: every time f is finished, it
     // is rescheduled to run deltaSeconds later. If you
     // need more accurate scheduling, don't use this method.
-    void scheduleEvery(Function f, int64_t deltaSeconds);
+    void scheduleEvery(Function f, int64_t deltaMilliSeconds);
```
Comment (random-zebra) [inline src/scheduler.h:55]: nit: comments need to be updated (they still mention the old argument `deltaSeconds`).

**Review by random-zebra — APPROVED** [review]
tested ACK 4bdf68aff1e5aec69302d55ed2239a7d81c28457

File: src/scheduler.h (hunk 2 above)
Comment (furszy) [reply src/scheduler.h]: done, squashed into the last commit.

**Review by random-zebra — APPROVED** [review]
utACK 98d770f818f848985523d4787dab148747cb858c

**Review by Fuzzbawls — APPROVED** [review]
utACK 98d770f818f848985523d4787dab148747cb858c

**Pull request merged**